Implement oauth2_auth_token

I don't know if any API additions need to be done to make this use case easy.

https://github.com/grpc/grpc/blob/master/doc/interop-test-descriptions.md#oauth2_auth_token

I never pushed an artifact before, so I'm wondering if it's normal that a newly pushed artifact can not be seen after 6 hours?
```
FAILURE: Build failed with an exception.

* What went wrong:
Could not resolve all dependencies for configuration ':grpc-auth:compile'.
> Could not find com.google.auth:google-auth-library-oauth2-http:0.2.0.
  Searched in the following locations:
      https://repo1.maven.org/maven2/com/google/auth/google-auth-library-oauth2-http/0.2.0/google-auth-library-oauth2-http-0.2.0.pom
      https://repo1.maven.org/maven2/com/google/auth/google-auth-library-oauth2-http/0.2.0/google-auth-library-oauth2-http-0.2.0.jar
      file:/usr/local/google/home/simonma/.m2/repository/com/google/auth/google-auth-library-oauth2-http/0.2.0/google-auth-library-oauth2-http-0.2.0.pom
      file:/usr/local/google/home/simonma/.m2/repository/com/google/auth/google-auth-library-oauth2-http/0.2.0/google-auth-library-oauth2-http-0.2.0.jar
  Required by:
      io.grpc:grpc-auth:0.8.0-SNAPSHOT

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output.

BUILD FAILED

```

--- a/build.gradle
+++ b/build.gradle
@@ -126,7 +126,7 @@ subprojects {
                 javaee_api: 'javax:javaee-api:7.0',
                 jsonp: 'org.glassfish:javax.json:1.0.4',
                 jsr305: 'com.google.code.findbugs:jsr305:3.0.0',
-                oauth_client: 'com.google.auth:google-auth-library-oauth2-http:0.1.0',
+                oauth_client: 'com.google.auth:google-auth-library-oauth2-http:0.2.0',
                 okhttp: 'com.squareup.okhttp:okhttp:2.4.0',
                 protobuf: "com.google.protobuf:protobuf-java:${protobufVersion}",
                 protobuf_nano: "com.google.protobuf.nano:protobuf-javanano:${protobufVersion}",

--- a/interop-testing/src/main/java/io/grpc/testing/integration/AbstractTransportTest.java
+++ b/interop-testing/src/main/java/io/grpc/testing/integration/AbstractTransportTest.java
@@ -43,8 +43,10 @@ import static org.mockito.Mockito.timeout;
 import static org.mockito.Mockito.verify;
 import static org.mockito.Mockito.verifyNoMoreInteractions;
 
+import com.google.auth.oauth2.AccessToken;
 import com.google.auth.oauth2.ComputeEngineCredentials;
 import com.google.auth.oauth2.GoogleCredentials;
+import com.google.auth.oauth2.OAuth2Credentials;
 import com.google.auth.oauth2.ServiceAccountCredentials;
 import com.google.protobuf.ByteString;
 import com.google.protobuf.EmptyProtos.Empty;
@@ -772,6 +774,41 @@ public abstract class AbstractTransportTest {
             .setBody(ByteString.copyFrom(new byte[314159])))
         .build();
     assertEquals(goldenResponse, response);
+  }
+
+  /** Sends a unary rpc with raw oauth2 access token credentials. */
+  public void oauth2AuthToken(String jsonKey, InputStream credentialsStream, String authScope)
+      throws Exception {
+    GoogleCredentials utilCredentials =
+        GoogleCredentials.fromStream(credentialsStream);
+    utilCredentials = utilCredentials.createScoped(Arrays.<String>asList(authScope));
+    AccessToken accessToken = utilCredentials.refreshAccessToken();
+
+    // TODO(madongfly): The Auth library may have something like AccessTokenCredentials in the
+    // future, change to the official implementation then.
+    OAuth2Credentials credentials = new OAuth2Credentials(accessToken) {
+      @Override
+      public AccessToken refreshAccessToken() throws IOException {
+        throw new IOException("This credential is based on a certain AccessToken, "
+            + "so you can not refresh AccessToken");
+      }
+    };
+
+    TestServiceGrpc.TestServiceBlockingStub stub = blockingStub
+        .withInterceptors(new ClientAuthInterceptor(credentials,
+            Executors.newSingleThreadExecutor()));
+    final SimpleRequest request = SimpleRequest.newBuilder()
+        .setFillUsername(true)
+        .setFillOauthScope(true)
+        .build();
+
+    final SimpleResponse response = stub.unaryCall(request);
+    assertFalse(response.getUsername().isEmpty());
+    assertTrue("Received username: " + response.getUsername(),
+        jsonKey.contains(response.getUsername()));
+    assertFalse(response.getOauthScope().isEmpty());
+    assertTrue("Received oauth scope: " + response.getOauthScope(),
+        authScope.contains(response.getOauthScope()));
   }
 
   protected static void assertSuccess(StreamRecorder<?> recorder) {

--- a/interop-testing/src/main/java/io/grpc/testing/integration/TestServiceClient.java
+++ b/interop-testing/src/main/java/io/grpc/testing/integration/TestServiceClient.java
@@ -158,6 +158,18 @@ public class TestServiceClient {
           + "\n                              Defaults to server host"
           + "\n  --server_port=PORT          Port to connect to. Default " + c.serverPort
           + "\n  --test_case=TESTCASE        Test case to run. Default " + c.testCase
+          + "\n    Valid options:"
+          + "\n      empty_unary: empty (zero bytes) request and response"
+          + "\n      large_unary: single request and (large) response"
+          + "\n      client_streaming: request streaming with single response"
+          + "\n      server_streaming: single request with response streaming"
+          + "\n      ping_pong: full-duplex ping-pong streaming"
+          + "\n      empty_stream: A stream that has zero-messages in both directions"
+          + "\n      service_account_creds: large_unary with service_account auth"
+          + "\n      compute_engine_creds: large_unary with compute engine auth"
+          + "\n      oauth2_auth_token: raw oauth2 access token auth"
+          + "\n      cancel_after_begin: cancel stream after starting it"
+          + "\n      cancel_after_first_response: cancel on first response"
           + "\n  --use_tls=true|false        Whether to use TLS. Default " + c.useTls
           + "\n  --use_test_ca=true|false    Whether to trust our fake CA. Default " + c.useTestCa
           + "\n  --use_okhttp=true|false     Whether to use OkHttp instead of Netty. Default "
@@ -211,16 +223,20 @@ public class TestServiceClient {
       tester.pingPong();
     } else if ("empty_stream".equals(testCase)) {
       tester.emptyStream();
-    } else if ("cancel_after_begin".equals(testCase)) {
-      tester.cancelAfterBegin();
-    } else if ("cancel_after_first_response".equals(testCase)) {
-      tester.cancelAfterFirstResponse();
     } else if ("compute_engine_creds".equals(testCase)) {
       tester.computeEngineCreds(defaultServiceAccount, oauthScope);
     } else if ("service_account_creds".equals(testCase)) {
       String jsonKey = Files.toString(new File(serviceAccountKeyFile), Charset.forName("UTF-8"));
       FileInputStream credentialsStream = new FileInputStream(new File(serviceAccountKeyFile));
       tester.serviceAccountCreds(jsonKey, credentialsStream, oauthScope);
+    } else if ("oauth2_auth_token".equals(testCase)) {
+      String jsonKey = Files.toString(new File(serviceAccountKeyFile), Charset.forName("UTF-8"));
+      FileInputStream credentialsStream = new FileInputStream(new File(serviceAccountKeyFile));
+      tester.oauth2AuthToken(jsonKey, credentialsStream, oauthScope);
+    } else if ("cancel_after_begin".equals(testCase)) {
+      tester.cancelAfterBegin();
+    } else if ("cancel_after_first_response".equals(testCase)) {
+      tester.cancelAfterFirstResponse();
     } else {
       throw new IllegalArgumentException("Unknown test case: " + testCase);
     }

